test/gtest/ucp/test_ucp_proto.cc

```diff
@@ -353,3 +358,170 @@ UCS_TEST_P(test_perf_node, replace_null)
+
+            for (int stage = 0; stage < UCP_PROTO_STAGE_LAST; ++stage) {
+                m_progress_storage[proto][stage] = proto->progress[stage];
+                proto->progress[stage]           = &progress_wrapper;
+            }
+        }
+    }
+
+    void protos_restore_progress()
+    {
+        for (int proto_id = 0; proto_id < ucp_protocols_count(); ++proto_id) {
+            auto proto = const_cast<ucp_proto_t*>(ucp_protocols[proto_id]);
+
+            for (int stage = 0; stage < UCP_PROTO_STAGE_LAST; ++stage) {
+                proto->progress[stage] = m_progress_storage[proto][stage];
+            }
+        }
+    }
+
+    void replace_ep_put_zcopy(ucp_ep_h ep) {
+        for (auto lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
+            if (lane == ucp_ep_get_cm_lane(ep)) {
+                continue;
+            }
+
+            auto iface  = ucp_ep_get_lane(ep, lane)->iface;
+            auto result = m_sender_uct_ops.emplace(iface, iface->ops);
+            if (result.second) {
+                iface->ops.ep_put_zcopy =
+                        reinterpret_cast<uct_ep_put_zcopy_func_t>(
+                                ucs_empty_function_return_unsupported);
+            }
+        }
+    }
+
+    static ucs_status_t progress_wrapper(uct_pending_req_t *uct_req)
+    {
+        ucs_status_t status;
+        auto req   = ucs_container_of(uct_req, ucp_request_t, send.uct);
+        auto proto = req->send.proto_config->proto;
+        auto stage = req->send.proto_stage;
+        std::string proto_name = req->send.proto_config->proto->name;
+
+        if (proto_name == "rndv/put/zcopy") {
+            if (stage == 0) {
+                if (!self->is_ep_closed) {
+                    scoped_log_handler slh(wrap_errors_logger);
+                    self->receiver().close_all_eps(*self, 0, UCP_EP_CLOSE_FLAG_FORCE);
+                    self->replace_ep_put_zcopy(req->send.ep);
+                    self->is_ep_closed = true;
+                }
+            } else {
+                EXPECT_TRUE(false) << "Unexpected proto stage";
+            }
+        }
+
+        status = self->m_progress_storage[proto][stage](uct_req);
+
+        if (self->is_ep_closed) {
+            for (auto &elem : self->m_sender_uct_ops) {
+                elem.first->ops = elem.second;
+            }
+            self->protos_restore_progress();
+        }
+
+        return status;
+    }
+
+    void rndv_put_abort_test()
+    {
+        /* Do the first send-receive loop to establish the connection
+         * and check that connection works.
+         */
+        ucp_request_param_t req_param{};
+        auto sreq = ucp_tag_send_nbx(sender().ep(), &m_send_buf.front(),
+                                     m_send_buf.size(), 0, &req_param);
+        auto rreq = ucp_tag_recv_nbx(receiver().worker(), &m_recv_buf.front(),
+                                     m_recv_buf.size(), 0, 0, &req_param);
+        requests_wait({sreq, rreq});
```

Comment:
seems like `test_ucp_peer_failure::smoke_test`

Reply:
I want to check exactly RNDV transfer and `test_ucp_peer_failure` doesn't check full rndv transfer because it has no `recv()` call during the am callback execution. I don't think it worth to implement this logic for `test_ucp_peer_failure` since it would complicate code that isn't use during the `test_ucp_peer_failure` test. So I prefer to leave the things as is.